implement stateless resets

marten-seemann opened the pull request:
This PR changes the server state machine such that sending HelloRetryRequests can be performed statelessly.

In non-blocking mode, `Handshake()` will return a new alert, `AlertStatelessRetry`. When receiving this return, an app can discard all mint-related state (i.e. delete `Conn` garbage collected). When not using non-blocking mode, the handshake is performed as stateless as possible, at no point the initial ClientHello and the HelloRetryRequest will be stored.

The cookie that is sent to the client in the HelloRetryRequest is composed of two parts: In the mint-part of the cookie, we store the hash of the initial ClientHello. The second part can be used to store an application-defined cookie (used when `Config.CookieHandler` is set). The application has to provide a way to encrypt this cookie (it should use authenticated encryption for this). This can be done by setting `Config.CookieSource`. In blocking mode, we default to use AES-GCM if no `CookieSource` is set.

@ekr, @bifurcation: Please tell me what you think about this. The stateless handshake works for me with this PR, but I can't guarantee I didn't miss anything subtle here. My current understanding of the TLS 1.3 draft doesn't go a lot farther than what I need for my mint changes and the QUIC integration.

ekr commented:
@marten-seemann this seems to fail in CI...

ekr commented on alert.go:
```diff
@@ -82,6 +83,7 @@ var alertText = map[Alert]string{
 	AlertUnknownPSKIdentity:          "unknown PSK identity",
 	AlertNoApplicationProtocol:       "no application protocol",
 	AlertNoRenegotiation:             "no renegotiation",
+	AlertStatelessRetry:              "stateless retry",
```
This isn't your fault, but I wonder if we should give these non-alerts a special name so we don't acccidentally try to send them.

ekr commented on conn.go:
```diff
@@ -74,9 +74,16 @@ type Config struct {
 	AllowEarlyData     bool
 	// Require the client to echo a cookie.
 	RequireCookie bool
+	// TODO: fix comment here: no more default cookie handler
```
This TODO seems like it needs fixing.

ekr commented on conn.go:
```diff
@@ -74,9 +74,16 @@ type Config struct {
+	// The CookieSource is used to encrypt / decrypt cookies.
+	// TODO: implement a check for that
+	// If non-blocking mode is used, and cookies are required, this field has to be set.
+	// TODO: implement that
```
These TODOs also seem to need fixing

ekr commented on conn.go:
```diff
@@ -642,6 +648,19 @@ func (c *Conn) HandshakeSetup() Alert {
+			var err error
+			caps.CookieSource, err = newDefaultCookieSource()
+			if err != nil {
+				logf(logTypeHandshake, "Error initializing client state: %v", alert)
```
Should this say server state?

ekr commented on conn.go:
```diff
@@ -687,6 +706,7 @@ func (c *Conn) Handshake() Alert {
 
 	for !connected {
 		// Read a handshake message
+		logf(logTypeHandshake, "readmessage")
```
Maybe make this |logTypeVerbose| and make the text clearer.

ekr commented on conn.go:
```diff
@@ -720,6 +738,13 @@ func (c *Conn) Handshake() Alert {
+		// 	if c.config.NonBlocking {
+		// 		return AlertStatelessRetry
+		// 	}
+		// }
```
You should remove this code.

ekr commented on cookie_source.go:
```diff
@@ -0,0 +1,73 @@
+		return nil, err
+	}
+	return key, nil
+}
```
I'm not sure it's necessary to have an HKDF stage here. You are feeding in random bits, right?

ekr commented on cookie_source.go:
```diff
@@ -0,0 +1,73 @@
+	if _, err := rand.Read(nonce); err != nil {
+		return nil, err
+	}
+	return s.aead.Seal(nonce, nonce, data, nil), nil
```
I don't think this is the crypto structure you want here. The issue is that you have a non-trivial chance of nonce-reuse at 2^96 because of birthday, which is catastrophic with GCM. You should either use an integer nonce (that won't repeat within any reasonable bound) or instead generate a longer nonce (256 bits is good) and use it to derive both the per-message key and per-message nonce

ekr commented on server-state-machine.go:
```diff
@@ -57,13 +60,18 @@ import (
+type cookie struct {
+	MintCookie        []byte `tls:"head=2"`
+	ApplicationCookie []byte `tls:"head=2"`
+}
```
In NSS, we also start the negotiated parameters and double-check that they don't change

ekr commented on server-state-machine.go:
```diff
@@ -129,9 +139,29 @@ func (state ServerStateStart) Next(hm *HandshakeMessage) (HandshakeState, []Hand
+		plainCookie, err := state.Caps.CookieSource.DecodeToken(clientCookie.Cookie)
+		if err != nil {
+			logf(logTypeHandshake, fmt.Sprintf("[ServerStateStart] Error decoding token [%v]", err))
+			return nil, nil, AlertAccessDenied
```
I don't think you want AccessDenied here, but rather DecryptError or IllegalParameter

ekr replied:
I would put a TODO here because if you were to change your key, this wouldn't work.

ekr commented on server-state-machine.go:
```diff
@@ -129,9 +139,29 @@ func (state ServerStateStart) Next(hm *HandshakeMessage) (HandshakeState, []Hand
+		if _, err := syntax.Unmarshal(plainCookie, cookie); err != nil {
+			logf(logTypeHandshake, fmt.Sprintf("[ServerStateStart] Error unmarshaling cookie [%v]", err))
+			return nil, nil, AlertAccessDenied
+		}
```
This should be internal_error, because it can't happen.

ekr commented on server-state-machine.go:
```diff
@@ -142,26 +172,51 @@ func (state ServerStateStart) Next(hm *HandshakeMessage) (HandshakeState, []Hand
-			logf(logTypeHandshake, "[ServerStateStart] Error computing truncated ClientHello [%v]", err)
-			return nil, nil, AlertDecodeError
+			logf(logTypeHandshake, "[ServerStateStart] Error in PSK negotiation [%v]", err)
+			return nil, nil, AlertInternalError
```
I am surprised this is internal error. @bifurcation ?

ekr commented on server-state-machine.go:
```diff
@@ -142,26 +172,51 @@ func (state ServerStateStart) Next(hm *HandshakeMessage) (HandshakeState, []Hand
+			binder := computeFinishedData(params, binderKey, ctxHash.Sum(nil))
+			if !bytes.Equal(binder, clientPSK.Binders[selectedPSK].Binder) {
+				logf(logTypeNegotiation, "Binder check failed for identity %x; [%x] != [%x]", psk.Identity, binder, clientPSK.Binders[selectedPSK].Binder)
+				return nil, nil, AlertInternalError
```
This seems like decrypt_error.

ekr commented on server-state-machine.go:
```diff
@@ -175,61 +230,66 @@ func (state ServerStateStart) Next(hm *HandshakeMessage) (HandshakeState, []Hand
+					logf(logTypeHandshake, "[ServerStateStart] Error generating cookie [%v]", err)
+					return nil, nil, AlertInternalError
+				}
+				shouldSendHRR = appCookie != nil
```
Is this idiomatic go? I would have done if....

marten-seemann commented:
Hi, @ekr thanks for the review! Seems like I forgot a bunch of TODOs and a few debug statements. I removed those, and added a bunch of documentation about `Cookie`s, `CookieSource`s and `CookieHandler`.
The reason the tests were failing is that CircleCI apparently uses Go 1.8 if no config file is provided, and the test helper functions (`t.Helper()`) require Go 1.9. I added a basic config file.

I really like the idea of storing the negotiated parameters in the cookie as well. By doing so, I was able to remove a rather ugly hack that was necessary to make PSK mode work. Is there anything else besides the `CipherSuite` that we would want to store in the cookie?

The code for the cookie source is what we copied from Google's QUIC implementation 1.5 years ago or so. I changed the cookie source to use a nonce size of 32 byte, as you suggested, and also removed the HKDF. I'm not sure though why're suggesting using a per-message key, I thought using the same key will be ok as long as the nonce is not reused. Can you please explain?
I also made the `DefaultCookieSource` public. When using non-blocking mode, an app will have to provide its own implementation for the `CookieSource`. The `DefaultCookieSource` should work fine in easy use cases, such as running a single server (in a multi-server deployment you'll need some shared cookie source anyway, and probably also implement some key rotation mechanism).

ekr commented:
The issue with the nonce here is that AES-GCM fails catastrophically if you have nonce reuse. If you use a random nonce, you get a non-trivial probability of reuse with enough connections (the 50% mark is at 2^{48} but the risk is unacceptably high well below that). Unfortunately, just making the nonce bigger doesn't help because AES-GCM hashes a >96 bit nonce down to 96 bits.

If you want to use GCM, you can guarantee a unique key/nonce pair by generating a longer nonce and then using HKDF to turn the master key into a key/nonce. Because the output is longer, the chance of collision is reduced.

marten-seemann commented:
I just pushed a fixed. We're now creating a new key (which is HKDF derived using the nonce as a salt) for every nonce.

ekr commented on cookie_source.go:
```diff
@@ -11,63 +11,68 @@ import (
 
-const tokenKeySize = 16
-const tokenNonceSize = 16
+var _ CookieSource = &DefaultCookieSource{}
```
I am probably not enough of a Golang wizard, but why is this neded.

marten-seemann replied:
Strictly speaking this is not needed. It makes the compiler check that `DefaultCookieSource` actually implements the `CookieSource` interface. Since it's an assignment to `_`, it will be removed during compilation.

ekr commented on cookie_source.go:
```diff
@@ -11,63 +11,68 @@ import (
+	if err != nil {
+		return nil, err
+	}
+	return aead.Seal(nonce, nonce, data, nil), nil
```
I don't think this is quite what you want. Rather, you want to do:

```
r := make([]byte, 32)
rand.Read(r)
key = hkdf(secret, R, []byte("mint cookie key")) // key length
nonce = hkdf(secret, R, []byte("mint cookie nonce")) // 12 bytes
```

And then you can use nonce the way you have it here.

ekr replied:
Or, if you prefer you can just read the key and the nonce as successive values from the hkdf.

marten-seemann replied:
How is that supposed to work? When decoding the cookie, I don't know `r` any more, and I can't create the AEAD to open the cookie.

ekr replied:
The cookie is r || aead.Seal(blah...)

Right now you are just using nonce == r, so the difference I am proposing is you hkdf r to make the nonce.

ekr commented on cookie_source.go:
```diff
@@ -11,63 +11,68 @@ import (
+	if err != nil {
+		return nil, err
+	}
+	return cipher.NewGCMWithNonceSize(c, cookieNonceSize)
```
You want NewGCM()

ekr commented on cookie_source_test.go:
```diff
@@ -0,0 +1,27 @@
+		_, err := cs.DecodeToken([]byte("too short"))
+		assertError(t, err, "it should reject too short tokens")
+		_, err = cs.DecodeToken(append(bytes.Repeat([]byte{0}, cookieNonceSize), []byte("invalid token")...))
+		assertError(t, err, "it should reject invalid tokens")
```
It would also be good to have one with a bogus nonce, e.g., by stomping the first byte.

marten-seemann commented:
@ekr, I implemented the changes for the `CookieSource` as you suggested. Please have another look.

ekr commented:
@marten-seemann I think you forgot to push

marten-seemann commented:
I don't think so. I rebased and amended the commit.

ekr commented:
Oh, ok. Will take a look

On Thu, Nov 30, 2017 at 1:44 PM, Marten Seemann <notifications@github.com>
wrote:

> I don't think so. I rebased and amended the commit.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bifurcation/mint/pull/141#issuecomment-348330529>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABD1oWBf6g5W4Xakvb7Mqyicm_25VN12ks5s7yGhgaJpZM4QVzSU>
> .
>

ekr commented on cookie_source.go:
```diff
@@ -0,0 +1,86 @@
+	aeadNonce := make([]byte, 12)
+	if _, err := io.ReadFull(h, aeadNonce); err != nil {
+		return nil, nil, err
+	}
```
This looks fine to me. Assuming you didn't change anything else, I think this is good to Go. @bifurcation ??

marten-seemann replied:
I didn't change anything else. This should be ready to merge now.

bifurcation commented on .circleci/config.yml:
```diff
@@ -0,0 +1,10 @@
+version: 2
```
Unless this change is necessary for this functionality, please put it in a separate PR.

marten-seemann replied:
This was the first commit of this PR (03b32d6dfee03aaed5b3759a24fe5ef8d54cd943). I openend #147, containing just this one commit.

marten-seemann commented:
@bifurcation: Is there anything else you need from me to merge this PR?

bifurcation commented on cookie_source.go:
```diff
@@ -0,0 +1,86 @@
+package mint
```
For consistency with the other files, this should be `cookie-source.go` (reserving underscores for `_test`).

bifurcation commented on conn.go:
```diff
@@ -74,9 +78,16 @@ type Config struct {
-	// The default cookie handler uses 32 random bytes as a cookie.
-	CookieHandler     CookieHandler
+	// A CookieHandler can be used to set and validate a cookie.
+	// The cookie returned by the CookieHandler will be part of the cookie sent on the wire, and endoced using the CookieSource.
```
Nit: "encoded"

bifurcation commented on conn.go:
```diff
@@ -74,9 +78,16 @@ type Config struct {
+	// It should make sure that the Cookie cannot be read and tampered with by the client.
+	// If non-blocking mode is used, and cookies are required, this field has to be set.
+	// In blocking mode, a default cookie source is used, if this is unused.
+	CookieSource      CookieSource
```
Nit: `CookieSource` doesn't seem like the right name for something whose job is to protect / unprotect cookie values.  `CookieProtection` or something like that?

marten-seemann commented:
@bifurcation: I rebased this PR on top of the current master, fixed the type, and renamed the `CookieSource` to `CookieProtector`.

bifurcation commented:
@marten-seemann Thanks! Merging and moving on to #145 ...
